Position.hpp

```diff
@@ -286,9 +286,13 @@ class Position {
    * counts number of bit set to one in a 64bits integer
    */
   static unsigned int popcount(position_t m) {
+    #if defined(__GNUC__) || defined(__GNUG__)
+    return __builtin_popcount(m);
+    #else
     unsigned int c = 0;
```

Comment:
For the specific usecase of this very spare (very few 1, a lot of 0) Connect4 bitmask, the current implementation is actually faster than the generic builtin implementation.

moreover, I want to avoid non C++ standard code as much as possible, the only exception I made was for 128bits integer that are not in C++ standard.

Reply:
Hey @PascalPons ,
I actually compiled both versions:
Yours:
```
popcount(int):
        push    rbp
        mov     rbp, rsp
        mov     DWORD PTR [rbp-20], edi
        mov     DWORD PTR [rbp-4], 0
        mov     DWORD PTR [rbp-4], 0
.L5:
        cmp     DWORD PTR [rbp-20], 0
        je      .L4
        mov     eax, DWORD PTR [rbp-20]
        sub     eax, 1
        and     DWORD PTR [rbp-20], eax
        add     DWORD PTR [rbp-4], 1
        jmp     .L5
.L4:
        mov     eax, DWORD PTR [rbp-4]
        pop     rbp
        ret
```
^ has this x86 output.

The builtin version has this output:
```
builtin_popcount(int):
        push    rbp
        mov     rbp, rsp
        sub     rsp, 16
        mov     DWORD PTR [rbp-4], edi
        mov     eax, DWORD PTR [rbp-4]
        mov     eax, eax
        mov     rdi, rax
        call    __popcountdi2
        leave
        ret
```

^Could you please give more details about your benchmarking? I'd love to know, as I enjoy optimization code (as this push probably implies ;) )

Reply:
I am using a benchmark data set of 6000 positions representative of different position complexity. See http://blog.gamesolver.org/solving-connect-four/02-test-protocol/ for the dataset.

I've tested the current implementation against the GCC built_ing popcount using GCC profiling with -O3 optimization. 

Note that you have to use __builtin_popcountll() for the 64bits version and not __builtin_popcount() as position_t is 64bits. Tou will have to continue using the current version when position_t is 128bits.

I also added the __attribute__ ((noinline)) to the popcount function in the profiling test code to ensure that the compiler does not inline it in order to be able to track its performance with gprof profiler.

the two version of the tested code looks are:
` 
static unsigned int __attribute__ ((noinline)) popcount(position_t m) {
    return __builtin_popcountll(m);

//    unsigned int c = 0;
//    for(c = 0; m; c++) m &= m - 1;
//    return c;
  }
`


here are the results:
**current version:**
Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  us/call  us/call  name    
 97.52     12.09    12.09    44675   270.68   275.83  GameSolver::Connect4::Solver::negamax(GameSolver::Connect4::Position const&, int, int)
  **1.86     12.32     0.23** 203288991     0.00     0.00 GameSolver::Connect4::Position::popcount(unsigned long)


**GCC builtin version:**
Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  us/call  us/call  name    
 97.43     12.02    12.02    44675   269.11   273.81  GameSolver::Connect4::Solver::negamax(GameSolver::Connect4::Position const&, int, int)
  **1.70     12.23     0.21** 203288991     0.00     0.00  GameSolver::Connect4::Position::popcount(unsigned long)


As you can see there is a small performance improvement of the builtin function (10% faster) on a function representing less than 2% of the total CPU time.

Overall, I do not feel this less than 0.2% performance gain is worth using a GCC specific function, moreover you also would have to continue to use the current version for the 128bits position_t as GCC does not provide a 128bits popcount. This would also lead to a less readable code

Pascal.

Reply:
Hey @PascalPons ,
Okay, fair. I was just trying to squeeze as much performance out as possible, but if you'd like, we can avoid that and use this.